Move media type categories to constants module

AetherUnbound opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #558 by @dhruvkb

Supersedes #560

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR uses the method described by @dhruvkb in #560 to move the media type categories to a constants module. In doing so, it also changes the category serializer field from a `CharField` to a `MultipleChoiceField`. This has some implications on how we convert the arguments into an Elasticsearch query, as we now need to account for the fact that the field's data is a `set` rather than a string that can be `split`.

I'm not sure this is the correct approach currently because we actually _lose_ some useful information in the validation error. Here's an example on `http://localhost:8000/v1/images/?q=wolf&category=sound`.

**Before**

```
HTTP 400 Bad Request
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "detail": {
        "category": [
            "Invalid category: sound. Available options: {'illustration', 'photograph', 'digitized_artwork'}"
        ]
    }
}
```

**After**

```
HTTP 400 Bad Request
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "detail": {
        "category": [
            "\"sound\" is not a valid choice."
        ]
    }
}

```

I don't know enough about DRF, but perhaps we might be able to alter the error message for category so it does display the acceptable values. I think this will require us to make a custom serializer field.

An _entirely different_ approach we could take is just to move the sets that are defined in the `validate_categories` functions on each serializer and move only that to the constants module. This would prevent us from having to handle the `set` type differently since it would still be a `CharField`, and we're already using the `_validate_enums` function elsewhere.

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
`just api-test` for unit tests

Use the link above to see the above results, then change the category to `photograph` or similar and see that results get returned.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

AetherUnbound commented:
Although now we don't need the help text, because it's generated for us:
![image](https://user-images.githubusercontent.com/10214785/160018686-9bc0df9d-8211-471a-a41a-3995747c297b.png)

obulat reviewed (commented):
TIL: the `CharField` and `MultipleChoiceField` have a different behavior when it comes to query parsing, and it's a breaking change here:
If you try to use two categories separated with commas(`http://localhost:8000/v1/images/?q=wolf&category=illustration,photograph`), you get an error:
```
HTTP 400 Bad Request
Allow: GET, HEAD, OPTIONS
Content-Type: application/json
Vary: Accept

{
    "detail": {
        "category": [
            "\"illustration,photograph\" is not a valid choice."
        ]
    }
}
```

The correct query for several `MultipleChoiceField`s would have each choice to have its own parameter:
`http://localhost:8000/v1/images/?q=wolf&category=illustration&category=photograph`.

So, my vote is for the _entirely different_ approach 😄

AetherUnbound commented:
Oof, I hadn't even considered that! I think changing this field type would pretty drastically affect the API. How do we feel about _just_ moving the list of strings that are valid categories into `constants/categories.py`?

zackkrida commented:
cc @dhruvkb and @obulat to discuss @AetherUnbound's follow-up question

zackkrida commented:
> How do we feel about just moving the list of strings that are valid categories into constants/categories.py?

Let's move forward with this, [as discussed](https://wordpress.slack.com/archives/C02012JB00N/p1649171676474039)!

AetherUnbound commented:
I've reverted most of the changes and moved the sets out to their own file. I also made a `Categories` subclass of `set` so we'd have a handy `make_help_text` function on it. This might be way overkill...but I wanted to make it so that our help text doesn't drift from the actual category set we have in code.

Lastly, I uncovered an issue that was introduced in #583 - we changed the category field from `categories` to `category`, but didn't change the `validate_categories` function to match. Unfortunately because some of this is DRF-magic, it wasn't immediately obvious that this meant the validation function for `category` was no longer being called. One of the test cases I added in `audio_integration_tests` caught that, so hopefully it won't happen again in the future (at least for this field). Yet another nudge from our repository to work on #557 I suppose :sweat_smile:

AetherUnbound commented:
Wow, the ordering doesn't seem deterministic! I'm sure it's because they're `set`s. This passed locally for me 😅 #WorksOnMyMachine

I'll fix that!

dhruvkb commented on api/catalog/api/constants/categories.py:
```diff
@@ -0,0 +1,35 @@
+class Categories(set):
+    def make_help_text(self) -> str:
```
This is partially duplicated here:

https://github.com/WordPress/openverse-api/blob/f2ff9ae525ae060b65608b08f34d6fbc193d4346/api/catalog/api/serializers/media_serializers.py#L68

dhruvkb replied:
Yours is a better implementation, might want to delete usage of this^ one.

AetherUnbound commented:
I've re-requested review because I've made some significant changes. Instead of having a specific `Categories` class, I've moved that help text functionality into its own module. I then applied that module everywhere that seemed reasonable in the serializers list. Here's a before (left) and after (right) on the help text:
![image](https://user-images.githubusercontent.com/10214785/163029007-5853638d-ef65-4463-a9b6-b08b6052a9f7.png)

AetherUnbound commented on api/catalog/api/serializers/image_serializers.py:
```diff
@@ -38,34 +44,33 @@ class ImageSearchRequestSerializer(
+        return value.lower()
+
+    @staticmethod
+    def validate_size(value):
+        _validate_enum("size", IMAGE_SIZES, value)
```
This validation wasn't happening before, so invalid `size` entries were permissible. This is now updated using the values defined in the help text.

dhruvkb commented:
The updated documentation is dope. 💯 &times; 💯  QoL improvement.

dhruvkb commented on api/catalog/api/constants/field_values.py:
```diff
@@ -0,0 +1,31 @@
+AUDIO_CATEGORIES = {
```
I think we should segregate the file by media types, or use `AUDIO_`/`IMAGE_` prefix before every variable. Otherwise it'll quickly become confusing as we add more media types.

AetherUnbound replied:
I was thinking about this...but there are some fields that might be relevant across media types. Like `DURATION` - I could see that being used for both audio _and_ video.

dhruvkb commented on api/catalog/api/serializers/audio_serializers.py:
```diff
@@ -62,35 +64,23 @@ class AudioSearchRequestSerializer(
-        help_text="A comma separated list of categories; available categories "
-        "include `music`, `sound_effect`, `podcast`, `audiobook`, "
-        "and `news`.",
+        help_text=make_comma_separated_help_text(AUDIO_CATEGORIES, "categories"),
```
This is fantastic, but a small part of me wants to extend `CharField`, give it a keyword argument (like `options` or `values`) and in `__init__`, generate this help text automatically. Again, this is clear and excellent and definitely a huge improvement **to merge as is**, but it is a bit repetitive.

AetherUnbound replied:
I started toying with this a bit; it gets a little more complicated...which is understandable, since we're turning more of this into "magic". I'd love to get this PR merged, then open a draft PR with an approach to this and see if folks would prefer that 🙂

dhruvkb commented on api/catalog/api/utils/help_text.py:
```diff
@@ -0,0 +1,16 @@
+
+def make_comma_separated_help_text(items: Iterable, name: str) -> str:
+    """
+    Generate help text that wraps each category in backticks.
```
We should document the params and the returned value.
```suggestion
    Generate help text that wraps each category in backticks.
    :param items: ...
    :param name: ...
    :return: ...
```

AetherUnbound replied:
Ah, and this also no longer describes `category` operations! I'll update this.

dhruvkb reviewed (approved):
Huge improvement over the present situation. I have some comments but they're not blocking.
